meta/state.go

```diff
@@ -0,0 +1,339 @@
+package meta
+
+import (
+	"errors"
+	"fmt"
+	"math/rand"
+	"net"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/hashicorp/raft"
+	"github.com/hashicorp/raft-boltdb"
+)
+
+// raftState abstracts the interaction of the raft consensus layer
+// across local or remote nodes.  It is a form of the state design pattern and allows
+// the meta.Store to change how its behavior with the raft layer at runtime.
+type raftState interface {
+	openRaft() error
+	initialize() error
+	leader() string
+	isLeader() bool
+	sync(index uint64, timeout time.Duration) error
+	setPeers(addrs []string) error
+	addPeer(addr string) error
+	invalidate() error
+	close() error
+	lastIndex() uint64
+	apply(b []byte) error
+	snapshot() error
+}
+
+// localRaft is a consensus strategy that uses a local raft implementation fo
+// consensus operations.
+type localRaft struct {
+	store     *Store
+	raft      *raft.Raft
+	transport *raft.NetworkTransport
+	peerStore raft.PeerStore
+	raftStore *raftboltdb.BoltStore
+	raftLayer *raftLayer
+}
+
+func (r *localRaft) invalidate() error {
+	time.Sleep(time.Second)
```

Comment:
Existing code that go moved over. Need to chat w/ @benbjohnson about this: https://github.com/influxdb/influxdb/blob/master/meta/store.go#L1265-L1266